tree_unique_args/src/loop_invariant.rs

```diff
@@ -0,0 +1,199 @@
+use std::iter;
+use strum::IntoEnumIterator;
+
+use crate::ir::{Constructor, Purpose, Sort};
+
+fn find_invariant_rule_for_ctor(ctor: Constructor) -> Option<String> {
+    let br = "\n      ";
+    let ruleset = " :ruleset always-run";
+
+    match ctor {
+        Constructor::Cons | Constructor::Nil | Constructor::Arg | Constructor::UnitExpr => None,
+        Constructor::Call => Some(format!(
```

Comment:
I think we should split out `find-in-Expr` into it's own analysis file, and rename it to `loop-child`

Reply:
Is that the same as BodyContains?

Reply:
This find-inv relation might be replaced by body-constain?

Reply:
Yeah, you could also do a more general `body-contains`